### ISSUE TITLE
Client side send-receive callback

### ISSUE THREAD

**Issue by niklasbjorkman** [Starcounter/Starcounter#3858]
Is there any callback in the client side framework where we can use to get a callback when the user invokes a call to the server and also when a response is received? We need a way to block the users from using the interface while a call i being processed server side (a "please wait" symbol).

We need to be able to differentiate the user induced calls from the Starcounter idle refresh that occurs without user interaction.

**Comment by miyconst** [comment]
The only callback is available is the JSON view-model state changes. In your case you can easily add a property `Busy$: false` to the client side. Set it to `true` on user interaction, and then revert back to `false` on server side once the calculations finished.

**Comment by niklasbjorkman** [comment]
Thanks for the reply. Yes, we could implement that but the problem is that we need some way to know when something is updating the JSON model in the client. Is there a way to do that without having to implement "busy" code manually on every component throughout the system. 

There are user interaction that is client side only, and do not update the model, and should therefor not set the busy flag. So we cannot make it on that level.

**Comment by dan31** [comment]
CC @tomalec @warpech

**Comment by miyconst** [comment]
@niklasbjorkman please look at the [Polymer observers](https://www.polymer-project.org/1.0/docs/devguide/observers).

**Comment by niklasbjorkman** [comment]
Thanks. We will try to implement that now client side. What we then need to do is to set the "Busy" to false server-side on each GET + handler. Can we somehow hook into all the responses centralized to prevent the requirement to remember to set Busy=false in all handler implementation?

**Comment by miyconst** [comment]
@niklasbjorkman there is an option to use middleware filter to catch all requests/responses. Please read this page - [Using middleware](http://starcounter.io/guides/network/using-middleware/).

**Comment by niklasbjorkman** [comment]
Neat. Implemented the Response middleware but the client never get a response and eventually chrome gives a timeout. The method is called and the "return response;" is reached. 

What am I doing wrong?
<code>
            Application.Current.Use((Request request, Response response) =>
            {
                if (Session.Current.Data != null)
                {
                    LoginSession ls = Session.Current.Data.Data as LoginSession;
                    if (ls != null)
                    {
                        ls.Busy = false;
                    }
                }
                return response;
            });

**Comment by niklasbjorkman** [comment]
Found one more thing to try. Please hold :)

**Comment by niklasbjorkman** [comment]
Ok, that solved the call to URI:s but how to handle the button click and value-changed events?

**Comment by miyconst** [comment]
```json
{
    "Html": "IndexPage.html",
    "LoadingButtonClick$": 0
}
```

```cs
public class IndexPage : Json {
    void Handle(Input.LoadingButtonClick Action) {
         Action.Cancel();
    }
}
```

```html
<button type="button" value="{{model.LoadingButtonClick$::click}}" disabled="{{model.LoadingButtonClick$}}" on-mouseup="++this.value">I am disabled on click until server responses</button>
```

**Comment by miyconst** [comment]
Explanation to the code above:

- There is a variable `LoadingButtonClick$` which is `0`, means `false`.
- The button applies `++` to the variable on click, and the value became `>0`, means `true`.
- The button has a `disabled="{{model.LoadingButtonClick$}}"` binding. Means the button will be disabled as soon as `LoadingButtonClick$` is `true`.
- On button click the `LoadingButtonClick$` became `true` and the button gets disabled.
- Then this change is being sent to the server side.
- The server cancels the action with `Action.Cancel();` and the value became `0` or `false` once again.
- The server sends this patch to the client.
- The client receives the patch and updates the view-model.
- The `LoadingButtonClick$` once again `0` or `true`.
- The button is no longer disabled.

In short:

>The button stays disabled until server responses to it's click. 

You can add any other logic on top of the `LoadingButtonClick$` variable.

**Comment by niklasbjorkman** [comment]
Hmm. A few things I do not get:
- This will only disable the button? We cannot trigger our "please wait" message that will lock the whole page for this?
- We actually need to Cancel() all Actions? 

Then it will be as much work as setting a .Busy = false manually on all requests?

**Comment by miyconst** [comment]
>This will only disable the button? We cannot trigger our "please wait" message that will lock the whole page for this?

As I already wrote, you can add any logic you needed on top of the `LoadingButtonClick$` variable. For example adding loading dialog.

```html
<template is="dom-if" if="{{model.LoadingButtonClick$}}">
    <div class="loading-dialog-box">Loading message here</div>
</template>
```

**Comment by niklasbjorkman** [comment]
Yes, but we can already achieve that globally with the suggested polymer observers? The problem is still the server side where we need to manually cancel all actions in all Handle() implementations. 

If we instead of using click-handlers use observers we do not need to handle click, evaluations and so on differently in the client if we use the Busy-flag proposed.

Resetting the Busy flag server side in the middleware can be done in one place for the Rest-URI-handlers.

So, only thing to automate is the reset of busy in the event Handles for all Page and Json implementation. Are there any way to do that?

**Comment by miyconst** [comment]
>So, only thing to automate is the reset of busy in the event Handles for all Page and Json implementation. 

Currently I can think of only one solution. Override the `HasChanged` function in your JSON class.

```cs
protected override void HasChanged(TValue property)
{
    this.Busy = false;
    base.HasChanged(property);
}
```

cc @chrhol
